### DISCUSSION TITLE
General questions about the operator

### DISCUSSION THREAD

**Discussion by awdonahue** [Opster/opensearch-k8s-operator#355] [Q&A] [answered]
My team and I are looking to move over to using Kubernetes to manage our Opensearch clusters, which inevitably led me to finding this operator. With that being the case I have started reviewing the documentation for our use cases and had a few questions.

1. I assume this is the case but would like to confirm that this operator in addition to EKS will work with core k8s and other open source options?

2. In the [development plan](https://github.com/Opster/opensearch-k8s-operator/blob/main/docs/designs/dev-plan.md) as part of _Completed Features_ a **scaling service** and under _Next Phases_ **automatic scaler** are mentioned. I assume the scaling service deals with scaling the opensearch cluster nodes up and down, but how is it controlling this? Is it CPU based, or memory, or something configurable? Also what is the difference between the **scaling service** and the **automatic scaler** that is upcoming?

3. The plan lists a monitoring service and built-in Prometheus exporter as a work in progress and it sounds great. With it being up and coming however, is there a recommended alternative to monitoring the internals of the Opensearch cluster? 

4. From a storage perspective, [this Opster article ](https://opster.com/guides/opensearch/opensearch-basics/set-up-run-opensearch-kubernetes-with-operator/) states "_The open-source Operator handles the attachment and detachment of nodes in a fully automatic process._". In the article it made it seem that before scaling down the operator makes sure to drain nodes preemptively to avoid the OS cluster from having unassigned shards which makes sense, but does the operator allow this behavior to be controlled? For example if I specify a single EBS disk for each of the nodes, can I move the disk assignments to different nodes directly or is draining the only supported option?

5. My organization primarily uses NGINX Ingress, however I was curious if there was a recommended ingress for use with the operator or if anything has seemingly performed better than others. I.E. Trafik, k8s Ingress, Istio, HAProxy, etc.

I hope this wasn't too long winded. I am going through the user-guide now and setting up a cluster with minikube. Thanks in advance for any answers!

**Comment by swoehrl-mw** [comment, marked as answer]
Hi @awdonahue. Great that your are considering using the operator.
To your questions:

1. The operator will work with any CNCF-certified kubernetes distribution (I know it has been deployed in EKS, AKS and K3s successfully).
2. In the current implementation you need to manually scale the cluster by changing the replicas in the nodepool definitions. The operator will then take care to scale the opensearch cluster accordingly (adding or removing nodes). It keeps the cluster state and health in mind and only removes one node at a time and waits for the cluster to recover (cluster health green) before removing the next node. The idea for the automatic scaler is to have the operator automatically add or remove nodes based on resource usage (not yet clear what, probably something like CPU, memory, disk).
3. You should be able to use any external exporter for elasticsearch/opensearch (https://github.com/prometheus-community/elasticsearch_exporter should work but haven't tested it myself)
4. Aside from implementing disk resizing the operator relies on the PVC handling of kubernetes and doesn't have any special support for reconfiguring or moving disks. It has an option that drains a node before restarting/updating it, but that is primarily designed for non-persistent disks.
5. You can use whatever ingress controller you like best or already use. I'm not aware of any performance specialities for any of them. Personally I also use nginx-ingress and it works great.

Hope this helps.

**Reply by awdonahue** [reply]
Thank you so much @swoehrl-mw, this helped immensely! We appreciate your time on your response. Cheers!
